### DISCUSSION TITLE
Removing Boundary Conditions from specific elements

### DISCUSSION THREAD

**Discussion by MatthewMavis** [idaholab/moose#16390] [Q&A Meshing] [answered]
Hi all,

I'm currently looking into trying to do topology optimisation using moose. I have looked into multiple ways of changing elements to do this. I have been able to create a meshgenerator to delete elements based on defined element IDs, though one problem with this is that to do so I'm using the ElementDeletionGeneratorBase.C which reassigns the element IDs after the elements have been removed. This means that the element IDs no longer match those of the original mesh which causes problems for example if you have to step back if you take off too much in the optimisation process. 

To get around this I have created another meshgenerator that instead changes the blockID of the element so that the element can have its density set to 0. The problem with this method is that the elements on the boundary condition interfaces retain the BCs. So is there any way to remove them on an element by element basis?

I'm also open to any ideas on how to retain the element IDs after deletion.

Best Regards,
Matt Mavis

**Comment by YaqiWang** [comment]
If you can change blockID of elements, why cannot you do block restriction to your variables, kernels, etc.? You can add new boundaries excluding the part connecting to a certain mesh block. I am not sure if this helps.

**Comment by permcody** [comment, marked as answer]
I'm not sure that MOOSE will support dynamically changing the boundary definition on-the-fly very well. However, there are a few workarounds to your issue:

1. Going back to your original issue, you could simply turn off the renumbering of elements. We do that to keep data structures contiguous, but this is not a requirement. There is an option in the `[Mesh]` block that will allow you to do this. 
2. There is a method in the BoundaryCondition class that lets you programmatically define when a boundary condition should be applied. You could extend the boundary condition you are using and define that method (`shouldApply`) to take into account your zero density property.
